File: Disjoint Set/16566번 - 카드 게임/16566.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int n, m, k;
vector<int> v;
int parent[4000001];

void input() {
    cin >> n >> m >> k;
    
	for(int i = 0, a; i < m; i++) {
		cin >> a;
		v.push_back(a);
		parent[i] = i;
	}
	
}

int Find(int x) {
	if(parent[x] == x) return x;
	return parent[x] = Find(parent[x]);
}

void Union(int a, int b) {
	a = Find(a);
	b = Find(b);
	if(a != b) parent[b] = a;
}

void solve() {
    sort(v.begin(), v.end());
	for(int i = 0, a; i < k; i++) {
		cin >> a;
		int idx = Find(upper_bound(v.begin(), v.end(), a) - v.begin());
		cout << v[idx] << '\n';
		Union(idx + 1, idx);
	}
}

int main()
{
	ios::sync_with_stdio(0);
	cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Disjoint Set/16566번 - 카드 게임/16566.py
import sys

input = sys.stdin.readline

n, m, k = map(int, input().split())
m_s = list(map(int, input().split()))
ch_s = list(map(int, input().split()))
parents = list(range(n + 1))

def Find(x):
    if x != parents[x]:
        parents[x] = Find(parents[x])
    return parents[x]
    
def Union(x, y):
    x = Find(x)
    y = Find(y)
    if x != y:
        parents[y] = x
        
def binarySearch(arr, k):
    st = 0
    end = len(arr) - 1
    while st <= end:
        mid = (st + end) // 2
        
        if k < arr[mid]:
            end = mid - 1
        else:
            st = mid + 1
    
    return st

def solve():
    m_s.sort()
    for ch in ch_s:
        idx = Find(binarySearch(m_s, ch))
        print(m_s[idx])
        Union(idx + 1, idx)

if __name__ == '__main__':
    solve()